### DISCUSSION TITLE
Argo Rollouts extensions

### DISCUSSION THREAD

**Discussion by spaghettifunk** [argoproj/argo-rollouts#2133] [Ideas]
Hey folks 👋 I love Argo Rollouts and I cannot thank you enough for this amazing project!
I have been working with Knative Serving for a while and I've been trying to figure out how I could combine Argo Rollouts and Knative together so that I could take advantage of all the goodies that Argo Rollouts has to offer and, at the same time, the simplicity of deploying applications with Knative Serving. I am aware of the discussion that happened [here](https://github.com/argoproj/argo-rollouts/issues/132#issuecomment-518519954) and I totally understand the issue and the developer experience that is at the base. 
This made me thinking: what if Argo Rollouts would have `extensions` like Argo CD? Would it be possible for Argo Rollout to control CRDs as well? How can Argo Rollouts "control" a generic CRD without knowing in advance what it is?

Where I am trying to get is basically the possibility to make available a set of `interfaces` that could be implemented by the external party that has the knowledge of the CRD. From Argo Rollouts perspective, the `Rollout` specification will have a reference to the code where the `interfaces` are implemented (like in ArgoCD) and the `workloadRef` would point to the CRD specification.
The `Controller` would now call the `interfaces` to execute the Rollout strategy. `Analysis` and `Experiments` shouldn't be touched by the external party since they are independent from the strategy itself. 

Does it make sense? 🙈

**Comment by kostis-codefresh** [comment]
@spaghettifunk Argo Rollouts has just released plugins in 1.5 https://argoproj.github.io/argo-rollouts/plugins/ (both for traffic providers and analysis providers)

> Where I am trying to get is basically the possibility to make available a set of interfaces that could be implemented by the external party that has the knowledge of the CRD

There is already a plugin for the Kubernetes Gateway API 

https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-gatewayapi

**Reply by spaghettifunk** [reply]
Hi @kostis-codefresh! This is awesome 🤩 I think now it would be possible to extend and use services like `KNative` with Argo Rollouts instead of their native way of doing traffic shifting! Thanks :)

**Reply by salaboy** [reply]
Yeah.. I think that by supporting the Gateway API we can integrate much easily with Knative, as there are efforts to support the Gateway API there too.

**Comment by lukasmrtvy** [comment]
Rollout Plugin [Proposal ](https://github.com/argoproj/argo-rollouts/blob/master/docs/proposals/resource-plugin.md?plain=1) and [PR ](https://github.com/argoproj/argo-rollouts/pull/4580)Draft
